components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1812,125 +1816,334 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
+                    + attributeValue);
+        }
+        // Check whether the filter operation is supported for filtering in groups.
+        if (isFilteringNotSupported(filterOperation)){
+            String errorMessage = "Filter operation: " + filterOperation + " is not supported for groups filtering.";
+            throw new CharonException(errorMessage);
         }
+        // Resolve the domain name in request according to isFilterUsersAndGroupsOnlyFromPrimaryDomainEnabled() or
+        // isFilteringEnhancementsEnabled() or domain name embedded in the filter attribute value.
+        domainName = resolveDomain(domainName, node);
+
         List<Object> filteredGroups = new ArrayList<>();
-        //0th index is to store total number of results;
+        // 0th index is to store total number of results.
         filteredGroups.add(0);
         try {
-            String[] roleList = getGroupList(attributeName, filterOperation, attributeValue);
+            String[] roleList = getGroupList(node, domainName);
             if (roleList != null) {
                 for (String roleName : roleList) {
                     if (roleName != null && carbonUM.isExistingRole(roleName, false)) {
-                        //skip internal roles
-                        if ((CarbonConstants.REGISTRY_ANONNYMOUS_ROLE_NAME.equals(roleName)) ||
-                                UserCoreUtil.isEveryoneRole(roleName, carbonUM.getRealmConfiguration())) {
+
+                        // Skip internal roles.
+                        if (CarbonConstants.REGISTRY_ANONNYMOUS_ROLE_NAME.equals(roleName) || UserCoreUtil
+                                .isEveryoneRole(roleName, carbonUM.getRealmConfiguration())) {
                             continue;
                         }
-                        String userStoreDomainName = IdentityUtil.extractDomainFromName(roleName);
-                        if (isInternalOrApplicationGroup(userStoreDomainName) || isSCIMEnabled(userStoreDomainName)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("SCIM is enabled for the user-store domain : " + userStoreDomainName + ". " +
-                                        "Including group with name : " + roleName + " in the response.");
-                            }
-                            Group group = getGroupWithName(roleName);
-                            filteredGroups.add(group);
-                        } else {
-                            if (log.isDebugEnabled()) {
-                                log.debug("SCIM is disabled for the user-store domain : " + userStoreDomainName + ". Hence " +
-                                        "group with name : " + roleName + " is excluded in the response.");
-                            }
-                        }
+                        getRoleWithAttributes(roleName, filteredGroups);
```

Comment:
Better to return only the Group instance from this method and to add it to the filteredGroups list within this scope

Reply:
Fixed: 788feb46892fd4c42062f6197f51d104401db991